content/docs/ai/ai-checkpoints.md

```diff
@@ -0,0 +1,342 @@
+
+This pattern works whether you manage Neon projects for your users (embedded approach like Replit) or users bring their own Neon accounts (BYON approach like v0).
+
+<Admonition type="tip" title="Quick guide">
+Use one root branch (main) as your permanent "active" branch with a stable connection string. Create snapshots for checkpoints, restore them with `finalize_restore: true` and `target_branch_id` set to your root branch ID, then poll operations until complete before connecting.
+</Admonition>
+
+## Why snapshots for checkpoints
+
+Neon offers two key primitives that could be used for checkpointing:
+
+**Branches** are designed for:
+
+- Development environments that need isolation
+- Testing features before production
+- Preview deployments with their own connection strings
+- Scenarios where you need a live, modifiable database
+
+**Snapshots** are designed for:
+
+- Point-in-time recovery
+- Creating restore points before risky operations
+- Maintaining a history of database states
+- Scenarios where you need immutable checkpoints
+
+### Why snapshots work better for agent checkpoints
+
+When using branches for checkpoints, you encounter:
+
+- **Dependency chains**: Parent branches cannot be deleted while child branches exist
+- **Connection string changes**: Each new branch has a different connection string
+- **Complex rollback**: Multiple API calls and connection updates required
+
+Snapshots solve these problems:
+
+- **Independent entities**: Any snapshot can be deleted at any time, in any order
+- **Stable connections**: With the active branch pattern, your connection string never changes
+- **Simple restore**: Single API call to restore any checkpoint
+- **Cost-efficient**: Snapshots charge only for volume of changes between checkpoints, matching the checkpoint pattern perfectly
+
+## Concepts and terminology
+
+### Neon core concepts
+
+- **Project**: Neon project that owns branches, computes, and snapshots
+- **Branch**: An isolated database environment with its own data and schema that you can connect to and modify
+- **Snapshot**: An immutable, point-in-time backup of a branch's schema and data. Read-only until restored - you can query snapshot data but cannot modify it directly
+- **Root Branch**: A branch with no parent (typically named `main` or `production`). Required for snapshot restores with `finalize_restore: true`
+- **Operations**: Backend operations that return operation IDs you must poll to completion
+
+### Pattern-specific terminology
+
+- **Active branch**: The single root branch your agent always connects to. Connection string never changes
+- **Preview branch**: Temporary branch created from a snapshot for safe exploration
+- **Meta database**: Separate database/table that stores checkpoint metadata (prompt text, `snapshot_id`, etc.). Keeps audit trail across restores
+
+## Core concept: The active branch pattern
```

Comment:
Interesting, so we don't recommend several different branches in agent-controlled projects? I feel like even with agents it makes sense to move to a dev & prod branch environment long term where the agent build in dev and deploy to prod. Is this not something we can model right now?